url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/server/SOCChannelList.java

```diff
@@ -0,0 +1,339 @@
+
+    /**
+     * constructor
+     */
+    public SOCChannelList()
+    {
+        channelMutexes = new Hashtable();
+        channelMembers = new Hashtable();
+        inUse = false;
+    }
+
+    /**
+     * take the monitor for this channel list
+     */
+    public synchronized void takeMonitor()
+    {
+        D.ebugPrintln("SOCChannelList : TAKE MONITOR");
+
+        while (inUse)
+        {
+            try
+            {
+                wait(1000);
+            }
+            catch (InterruptedException e)
+            {
+                System.out.println("EXCEPTION IN takeMonitor() -- " + e);
+            }
+        }
+
+        inUse = true;
+    }
+
+    /**
+     * release the monitor for this channel list
+     */
+    public synchronized void releaseMonitor()
+    {
+        D.ebugPrintln("SOCChannelList : RELEASE MONITOR");
+        inUse = false;
+        this.notify();
+    }
+
+    /**
+     * take the monitor for this channel
+     *
+     * @param channel  the name of the channel
+     * @return false   if the channel has no mutex
+     */
+    public boolean takeMonitorForChannel(String channel)
+    {
+        D.ebugPrintln("SOCChannelList : TAKE MONITOR FOR " + channel);
+
+        MutexFlag mutex = (MutexFlag) channelMutexes.get(channel);
+
+        if (mutex == null)
+        {
+            return false;
+        }
+
+        boolean done = false;
+
+        while (!done)
+        {
+            mutex = (MutexFlag) channelMutexes.get(channel);
+
+            if (mutex == null)
+            {
+                return false;
+            }
+
+            synchronized (mutex)
+            {
+                if (mutex.getState() == true)
+                {
+                    try
+                    {
+                        mutex.wait(1000);
+                    }
+                    catch (InterruptedException e)
+                    {
+                        System.out.println("EXCEPTION IN takeMonitor() -- " + e);
+                    }
+                }
+                else
+                {
+                    done = true;
+                }
+            }
+        }
+
+        mutex.setState(true);
+
+        return true;
+    }
+
+    /**
+     * release the monitor for this channel
+     *
+     * @param channel  the name of the channel
+     * @return false if the channel has no mutex
+     */
+    public boolean releaseMonitorForChannel(String channel)
+    {
+        D.ebugPrintln("SOCChannelList : RELEASE MONITOR FOR " + channel);
+
+        MutexFlag mutex = (MutexFlag) channelMutexes.get(channel);
+
+        if (mutex == null)
+        {
+            return false;
+        }
+
+        synchronized (mutex)
+        {
+            mutex.setState(false);
+            mutex.notify();
+        }
+
+        return true;
+    }
+
+    /**
+     * @return an enumeration of channel names (Strings)
+     */
+    public Enumeration getChannels()
+    {
+        return channelMembers.keys();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **178**. At this line the following field is not protected by synchronization: `channelMembers`. Your code typically protects this field with synchronization on object: `SOCChannelList.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `SOCChannelList.this`. For example, like at lines: **187**, **260**, and **287**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **187**, **260**, **287**, **314**, and **326**.